server/src/main/java/org/elasticsearch/action/support/replication/PendingReplicationActions.java

```diff
@@ -0,0 +1,98 @@
+import org.elasticsearch.index.shard.ShardId;
+import org.elasticsearch.threadpool.ThreadPool;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Map;
+import java.util.Set;
+import java.util.function.Consumer;
+
+public class PendingReplicationActions implements Consumer<ReplicationGroup>, Releasable {
+
+    private final Map<String, Set<RetryableAction<?>>> onGoingReplicationActions = ConcurrentCollections.newConcurrentMap();
+    private final ShardId shardId;
+    private final ThreadPool threadPool;
+
+    public PendingReplicationActions(ShardId shardId, ThreadPool threadPool) {
+        this.shardId = shardId;
+        this.threadPool = threadPool;
+    }
+
+    public void addPendingAction(String allocationId, RetryableAction<?> replicationAction) {
+        Set<RetryableAction<?>> ongoingActionsOnNode = onGoingReplicationActions.get(allocationId);
+        if (ongoingActionsOnNode != null) {
+            ongoingActionsOnNode.add(replicationAction);
+            if (onGoingReplicationActions.containsKey(allocationId) == false) {
+                replicationAction.cancel(new UnavailableShardsException(shardId,
+                    "Replica unavailable - replica could have left ReplicationGroup or IndexShard might have closed"));
+            }
+        } else {
+            replicationAction.cancel(new UnavailableShardsException(shardId,
+                "Replica unavailable - replica could have left ReplicationGroup or IndexShard might have closed"));
+        }
+    }
+
+    public void removeReplicationAction(String allocationId, RetryableAction<?> action) {
+        Set<RetryableAction<?>> ongoingActionsOnNode = onGoingReplicationActions.get(allocationId);
+        if (ongoingActionsOnNode != null) {
+            ongoingActionsOnNode.remove(action);
+        }
+    }
+
+    @Override
+    public synchronized void accept(ReplicationGroup replicationGroup) {
+        Set<String> trackedAllocationIds = replicationGroup.getTrackedAllocationIds();
+
+        for (String targetAllocationId : trackedAllocationIds) {
+            onGoingReplicationActions.putIfAbsent(targetAllocationId, ConcurrentCollections.newConcurrentSet());
+        }
+        ArrayList<Set<RetryableAction<?>>> toCancel = new ArrayList<>();
+        for (String allocationId : onGoingReplicationActions.keySet()) {
+            if (trackedAllocationIds.contains(allocationId) == false) {
+                toCancel.add(onGoingReplicationActions.remove(allocationId));
+            }
+        }
+
+        threadPool.executor(ThreadPool.Names.GENERIC).execute(() -> toCancel.stream()
+            .flatMap(Collection::stream)
+            .forEach(action -> action.cancel(new UnavailableShardsException(shardId, "Replica left ReplicationGroup"))));
+    }
+
+
+    @Override
+    public void close() {
+        ArrayList<Set<RetryableAction<?>>> toCancel = new ArrayList<>(onGoingReplicationActions.values());
+        onGoingReplicationActions.clear();
```

Comment:
This is only eventually consistent (and can race any of the other methods). It's not strictly a problem for the use case where we're using this, but I wonder if we should add a `recheck` method in this class that we call whenever we reschedule that makes sure that pendingReplicationActions has not been closed, or if we should properly serialize all access to the top-level keyset, e.g. using a read-write-lock.

Reply:
I don't understand the concern here.

Only `close` and `accept` can write to the `allocation_id->set` map. And they are synchronized. `addPendingAction` does not write to the to that map. It only writes to the set. And it checks to ensure that the set is still part of the `allocation_id->set` map after writing. And `close` and `accept` only access the keyset after removing from `allocation_id->set`.